Immutable ipfs-log

haadcode opened the pull request:
This PR is a re-write of the log code base to make it immutable. Originally discussed in #97.

Closes #97.
Closes #61.
Closes #98.
Closes #4.
Closes #99.
Closes #98.
Closes #64.

haadcode commented:
Also note that this will completely change the public API and there's no backwards compatibility, and generally this is a massive change.

haadcode commented:
@dignifiedquire Since you have experience with immutable modules, would it be possible for you to take a look at this and go through and see what you think? 

I'm looking for feedback on how you'd feel as a user to use this API, but also on implementation details if  you see something to comment about.

I'm not looking for feedback yet on code style nor on the documentation (WIP).

dignifiedquire commented:
@haadcode sure thing, will look through it in the coming days

haadcode commented on README.md:
```diff
@@ -325,6 +324,13 @@ Starting benchmark...
+- Immutable
+- O(1) writes, O(1) reads, O(n) joins
+- Joins can be made constant-time by limiting the join (eg. { limit: 1000 })
+- Performance target: > 60 ops/s for joins, > 60 ops/s for writes
```
Reminder: These should be increased to 90 ops/s

dignifiedquire commented:
@haadcode I hadn't time to look at this yet. I will have time for it next week though :)

dignifiedquire commented:
@haadcode it would be great to have source maps + non-minified builds for the browser for profiling and debugging

dignifiedquire commented on src/log.js:
```diff
@@ -1,153 +1,389 @@
+        const fetch = (hash) => LogUtils._fetchRecursive(ipfs, hash, all, amount, depth + 1, entry, onProgressCallback)
+        return mapSeries(entry.next, fetch, { concurrency: 1 })
+          .then((res) => res.concat([entry]))
+          .then((res) => res.reduce((a, b) => a.concat(b), [])) // flatten the array
```
it's probably more efficient to not generate a new array on every iteration, you can pull in `lodash.flatten` which is small and does it pretty efficiently.

haadcode commented:
@dignifiedquire added source maps. I can profile and debug the source code with these source maps even as a minified build. Let me know if you have any problems doing the same.

dignifiedquire commented on src/log.js:
```diff
@@ -1,153 +1,389 @@
+   * @returns {Promise<Log>}
+   */
+  static fromMultihash(ipfs, hash, length = -1, onProgressCallback) {
+    if (!ipfs) throw new Error('Ipfs instance not defined')
```
You are defining the ipfs missing error all over the place, maybe create one and reuse it?

dignifiedquire commented on src/log.js:
```diff
@@ -1,153 +1,389 @@
+    if (!hash) throw new Error('Invalid hash: ' + hash)
+    let logData
+    return ipfs.object.get(hash, { enc: 'base58' })
+      .then((res) => logData = JSON.parse(res.toJSON().data))
```
`logData` is only used in the next promise below, you could just return it rather than pull it out of the scope

haadcode replied:
Great catch!

dignifiedquire commented on src/log.js:
```diff
@@ -1,153 +1,389 @@
+  static toMultihash(ipfs, log) {
+    if (!ipfs) throw new Error('Ipfs instance not defined')
+    if (!log) throw new Error('Log instance not defined')
+    if (log.items.length < 1) throw new Error(`Can't serialize an empty log`)
```
why ` for a string?

haadcode replied:
'Can\'t serialize an empty log' looks ugly :)

dignifiedquire commented on src/log.js:
```diff
@@ -1,153 +1,389 @@
+    return Entry.create(ipfs, data, log.heads)
+      .then((entry) => {
+        // Add the entry to the previous log entries
+        const items = log.items.slice().concat([entry])
```
No need for `slice`, `concat` already creates a new array

dignifiedquire commented on src/log.js:
```diff
@@ -1,153 +1,389 @@
+      return log.heads
+      .map((e) => log.get(e))
+      .filter((e) => e !== undefined)
+      .slice()
```
no need for `slice`, `filter` already creates a new array

dignifiedquire commented on src/log.js:
```diff
@@ -1,153 +1,389 @@
+    const log2 = isFirst ? b : a
+
+    // Cap the size of the entries
+    const newEntries = log2.items.slice(0, size)
```
maybe check for `size` equals length to avoid a copy when none is needed

haadcode replied:
Not sure what you mean exactly. In any case, we need to create a copy here in order to not mutate the previous items/array.

dignifiedquire replied:
I missed that, ignore me

dignifiedquire commented on src/log.js:
```diff
@@ -1,153 +1,389 @@
+   * @returns {boolean}
+   */
+  static _isReferencedInChain(log, entry) {
+    return log.items.slice().reverse().find((e) => Entry.hasChild(e, entry)) !== undefined
```
no need for `reverse` or `slice`, as you only care if one exists, not in which order

haadcode replied:
This is an optimization, start from the latest (reversed array), ie. "warm" entries, so for example when the `entry` happens to be the latest entry added to the array, we find it without going through the whole array.

dignifiedquire replied:
Interesting, would be curious if that actually helps in perf, because the copy + reverse is not cheap for large arrays I expect. 

In either case might be good to leave a comment, so the next person changing the code knows about this.

haadcode replied:
Should benchmark this at some point again. I remember benchmarking it long time ago to reach that set of calls but a lot has changed then.

haadcode commented:
@dignifiedquire if you run the benchmarks, I'd be curious to know what kind of max numbers (per second and per 10s) you're seeing on your machine.

haadcode commented on src/entry.js:
```diff
@@ -56,19 +76,32 @@ module.exports = class Entry {
+   * @alias compare
+   */
+  static isEqual(a, b) {
+    return a.hash === b.hash
```
Gonna rework this and drop `compare` altogether and just keep the equality check.

haadcode commented on src/log.js:
```diff
@@ -1,153 +1,389 @@
+   * @param {boolean} [withHash=false] - If set to 'true', the hash of the entry is included
+   * @returns {string}
+   */
+  toString(withHash) {
```
You're absolutely right. Gonna remove the flag as this functionality was for debugging initially. A nicely formatted string is all that is needed :)

haadcode commented:
Thank you @dignifiedquire for reviewing this! Many things were caught and fixed/improved. I've incorporated your feedback to this PR for the parts that make sense. 

I've also merged the benchmarks you PRed. We'll have to work on those in the future as the ones I've had here are providing a bit more valuable information as to what's happening and what the numbers are. However, the benchmarks from your PR are good enough and there's no reason to not merge them in.

As for the style related feedback, I really do understand where you're coming from, but will not change the style for now. If/when we move this repo under IPFS org, we can convert the style to follow the guidelines you have for `js-ipfs`.

That all being said, I'm ready to merge this into master as soon as I have updated the documentation to reflect the API changes. Will hopefully get to it today.

dignifiedquire commented:
the class removal is not about style, but about overhead. I suspect that most engines will generate more memory allocations for a class vs a a simple collection of methods.

haadcode commented:
> the class removal is not about style, but about overhead. I suspect that most engines will generate more memory allocations for a class vs a a simple collection of methods.

Would you have some material to read up on this? My gut feeling is, given how frikkin smart the VMs are, is that they would optimize class w/ static methods vs. functions to be roughly the same. Also, I don't expect the log to be used in high volumes (many instances), but that might be the case for entry. That said, I would prefer to approach this like any performance optimization: measure and see if there's a problem. I quickly checked mem usage and allocation behaviour with the profiler and didn't see anything glaring.

dignifiedquire commented:
This is only testing classes with regular methods on them, but it varies generally: https://kpdecker.github.io/six-speed/

The reason why I bring it up, is that there was mention in a couple of articles that es6 features are usually not yet optimisable fully by all jits, which is for example the reason I have to use `var` instead of `let` and `const` in hot loops in borc. 

Regarding perf I agree we should probably try and see if there is any measurable impact

haadcode commented:
Having gone another round of refactoring, this is now ready for merge again. I've updated pretty much everything since the last CR, squashed the commits and rebased on master. 

@dignifiedquire if you want, another round of feedback would be highly appreciated. That said, I plan to merge to asap (today!), so if you feel there's something that really needs to change, let me know asap.

haadcode commented:
This is finally done! \o/ 🎆 Thank you everyone who helped!
